### ISSUE TITLE
The parent definition "fos_http_cache.test.client.abstract" ... does not exist.

### ISSUE THREAD

**Issue by dantleech** [FriendsOfSymfony/FOSHttpCacheBundle#238]
````
The parent definition "fos_http_cache.test.client.abstract" defined for definition "fos_http_cache.test.client.varnish" does not exist.
````

The `fos_http_cache.test.client.varnish` service extends this service, but that service is only loaded when the `test` definitions are loaded (i.e. when the `test` key is included in the config).

**Comment by dantleech** [comment]
nm. looks like I was using a 1.1 version.

**Comment by dantleech** [comment]
still get this in 1.3

**Comment by ddeboer** [comment]
Well, the `fos_http_cache.test.client.varnish` service is defined as [abstract](https://github.com/FriendsOfSymfony/FOSHttpCacheBundle/blob/master/Resources/config/varnish.xml#L21). 

When the `test` config key is found, the following [happens](https://github.com/FriendsOfSymfony/FOSHttpCacheBundle/blob/master/DependencyInjection/FOSHttpCacheExtension.php#L325):
* `test_client.xml` is loaded, which adds `fos_http_cache.test.client.abstract` to the container
* `fos_http_cache.test.client.varnish` is made concrete.

This way, `fos_http_cache.test.client.varnish`, abstract by default, can have a non-existing service as a parent (in case there’s no `test` key in the config).

What I don’t understand, is how you get that error. :wink: Do you have some dependency on `fos_http_cache.test.client.varnish` somewhere? When do you get the error?

**Comment by stof** [comment]
Well, why not defining the ``fos_http_cache.test.client.varnish`` only when it is needed (or remove it when it is not) ? this would prevent someone from extending it in their own service and then getting an error about the bad parent.

**Comment by dantleech** [comment]
ok. I was using `dev-master` of `symfony/symfony`. "downgrading" to 2.7 fixes this issue. Validation for the DI config also didn't seem to work with `dev-master`.

**Comment by ddeboer** [comment]
> why not defining the fos_http_cache.test.client.varnish only when it is needed (or remove it when it is not)

Our idea was to configure as many services as possible in the XML, so not construct them dynamically. But removing it when not needed is a good way to prevent errors.

> I was using dev-master of symfony/symfony

Good to know that there’s a problem there! We don’t test against 2.8-dev yet on Travis, so I’ll add that and find out what we need to change. For now, glad you fixed your problem, @dantleech.

**Comment by dbu** [comment]
ah, interesting. so symfony 2.8 is doing more validation? yeah lets
remove those services when tests are not configured. or move them to a
separate file that we only load when tests are configured. we might even
want to do that in the 1.x version to have it fixed when sf 2.8 comes
out. our 2.0 might take a moment to get out, e.g. until php-http is stable.

**Comment by ddeboer** [comment]
Fix tagged as 1.3.4.
